pkg/providers/github.go

```diff
@@ -0,0 +1,182 @@
+package providers
+
+import (
+	"context"
+	crypto_rand "crypto/rand"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/google/go-github/v43/github"
+	"github.com/spectralops/teller/pkg/core"
+	"golang.org/x/crypto/nacl/box"
+	"golang.org/x/oauth2"
+)
+
+const (
+	gitHubSplitPathCount = 2
+)
+
+// GitHubActionClient describe the GitHub action client
+type GitHubActionClient interface {
+	GetRepoPublicKey(ctx context.Context, owner, repo string) (*github.PublicKey, *github.Response, error)
+	CreateOrUpdateRepoSecret(ctx context.Context, owner, repo string, eSecret *github.EncryptedSecret) (*github.Response, error)
+	DeleteRepoSecret(ctx context.Context, owner, repo, name string) (*github.Response, error)
+	ListRepoSecrets(ctx context.Context, owner, repo string, opts *github.ListOptions) (*github.Secrets, *github.Response, error)
+}
+
+type GitHub struct {
+	clientActions GitHubActionClient
+}
+
+// NewGitHub create new GitHub provider
+func NewGitHub() (core.Provider, error) {
+
+	token := os.Getenv("GITHUB_AUTH_TOKEN")
+	if token == "" {
+		return nil, errors.New("missing `GITHUB_AUTH_TOKEN`")
+	}
+
+	ts := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: token},
+	)
+	tc := oauth2.NewClient(context.TODO(), ts)
+	client := github.NewClient(tc)
+
+	return &GitHub{clientActions: client.Actions}, nil
+}
+
+func (g *GitHub) Name() string {
+	return "GitHub"
+}
+
+func (g *GitHub) Put(p core.KeyPath, val string) error {
+
+	owner, repoName, err := g.parsePathToOwnerAndRepo(p)
+	if err != nil {
+		return err
+	}
+
+	publicKey, _, err := g.getRepoPublicKey(owner, repoName)
+	if err != nil {
+		return err
+	}
+
+	encryptedSecret, err := g.encryptSecretWithPublicKey(publicKey, p.Env, val)
+	if err != nil {
+		return err
+	}
+
+	_, err = g.createOrUpdateRepoSecret(context.TODO(), owner, repoName, encryptedSecret)
+
+	return err
+}
+
+func (g *GitHub) PutMapping(p core.KeyPath, m map[string]string) error {
+	for k, v := range m {
+		ap := p.WithEnv(k)
+		err := g.Put(ap, v)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (g *GitHub) GetMapping(p core.KeyPath) ([]core.EnvEntry, error) {
+	return nil, fmt.Errorf("does not supported by the %s provider", g.Name())
+}
+
+func (g *GitHub) Get(p core.KeyPath) (*core.EnvEntry, error) {
+	return nil, fmt.Errorf("does not supported by the %s provider", g.Name())
+}
+
+func (g *GitHub) Delete(p core.KeyPath) error {
+	owner, repoName, err := g.parsePathToOwnerAndRepo(p)
+	if err != nil {
+		return err
+	}
+
+	_, err = g.deleteRepoSecret(context.TODO(), owner, repoName, p.Env)
+	return err
+}
+
+func (g *GitHub) DeleteMapping(p core.KeyPath) error {
+
+	owner, repoName, err := g.parsePathToOwnerAndRepo(p)
+	if err != nil {
+		return err
+	}
+
+	opt := github.ListOptions{PerPage: 100}
```

Comment:
Is there automatic pagination here or just 1st page?
IMHO if pagination becomes too tricky, we can say that we support only up to 100 entries.

Usually I pick github clients by their ability to automatically paginate or if they have some convenient sample for generically paginating. Sometimes this is tricky because it means many requests in succession, so errors, throttling etc. can be an issue.

Reply:
Looks like the there is not option to use `auto pagination/getAll`, I can  use pagination like in there example: https://github.com/google/go-github#pagination

but i don't think that repo will have more then 100 secrets. 
WDYT?

Reply:
exactly what you said. 
so we can go with 100 just document it somehow

Reply:
done